tests/test_js.py

```diff
@@ -70,3 +70,15 @@ def test_JsonFetcher_decodes_waveform_events_correctly(
     json_fetcher._fetch_data = mock.MagicMock(return_value=mock_response)
     aa_data = json_fetcher.get_values(dummy_pv, EARLY_DATE, LATE_DATE)
     assert aa_data == data_2d_2_events
+
+
+def test_JsonFetcher_decodes_enum_events_correctly(
+    dummy_pv, json_fetcher,
+):
+    enum_json = utils.load_from_file("enum_event.json")
+    mock_response = utils.mock_response(json_str=enum_json)
+    json_fetcher._fetch_data = mock.MagicMock(return_value=mock_response)
+    aa_data = json_fetcher.get_values(dummy_pv, EARLY_DATE, LATE_DATE)
+    assert aa_data.enum_options[5] == "Special"
+    assert aa_data.enum_strings[0] == "User"
+    assert aa_data.values[0] == 4
```

Comment:
`assert aa_data[0].enum_string[0] == "User"` fails here

Reply:
Thanks for the review!

> 
> 
> It's a bit of an interesting API question. Cothread does a similar thing, as do other EPICS libraries, but they don't usually have arrays of events.

Good point, I didn't think to look at what other libraries do.

> 
> I have commented on a couple of details, but the interesting questions are:
> 
>     * if enum values are not available, do you think an array of empty strings is preferable to `None`?

Probably not from the point of view of the user, but selfishly it was more convenient for me, as type is consistent and you don't have to check that it is not None every time you want to access it.

>     * could we store enum options only on an `ArchiveEvent`?

Yes, and have a method to look up the enum string for the value dynamically. This would probably be more robust as you won't lose any information when getting the event out of the ArchiveData. You're unlikely to change the value but you might want to know what the other options were.

>     * as we store single values as arrays, we probably do need to provide enum_string as an array even though this is unlikely ever to have more than one value

Yes I couldn't think of a use-case that has a waveform of enums, but in principle it is possible I think.

Reply:
Oh whoops this comment is in the wrong place but never mind